app/src/main/java/com/vocabulary/myvocabulary/ui/dictionaries/ShareDictionaryViewModel.kt

```diff
@@ -0,0 +1,159 @@
+                    importedDictionaryDetails.postValue(Event(DictionaryDetails(t, dictionary.dictionaryName)))
+                }
+    }
+
+    fun fetchCsvUri(): Uri? = shareDictionaryRepository.getCsvData()
+
+    fun parseDataAndCreateWords(dictionaryId: Long, context: Context) {
+        if(shareDictionaryRepository.getCsvData() != null) {
+            val scheme = shareDictionaryRepository.getCsvData()!!.scheme
+            if (ContentResolver.SCHEME_CONTENT == scheme) {
+                try {
+                    val contentResolver = context.contentResolver
+                    val inputStream = contentResolver.openInputStream(shareDictionaryRepository.getCsvData()!!)
+
+                    val reader = BufferedReader(InputStreamReader(inputStream))
+                    val csvParser = CSVParser(reader, CSVFormat.DEFAULT)
+
+                    for (csvRecord in csvParser) {
+                        insertWordToDatabase(Word(
+                                containerDictionaryId = dictionaryId,
+                                translation = csvRecord.get(0),
+                                word = csvRecord.get(1),
+                                created = Calendar.getInstance().time
+                        ))
+                    }
+                } catch (ex: Exception) {
+                    ex.printStackTrace()
+                }
+            }
+
+        }
+
+    }
+
+    private fun insertWordToDatabase(word: Word) {
+        disposables += Completable.fromCallable { wordRepository.createWord(word) }
+                .subscribeOn(rxSchedulers.io())
+                .observeOn(rxSchedulers.main())
+                .subscribe()
+    }
+
+    private fun observeIsImport() {
+        disposables += shareDictionaryRepository.getImport()
+                .subscribeOn(rxSchedulers.io())
+                .observeOn(rxSchedulers.main())
+                .subscribe{ isImport.postValue(it) }
+    }
+
+    fun setIsImport(isImport: Boolean) {
+        shareDictionaryRepository.setIsImport(isImport)
+    }
+
+    fun saveCsvData(csv: Uri) {
+        shareDictionaryRepository.storeCsvData(csv)
+    }
+
+    private fun writeCsvFile(words: List<Word>, context: Context): File {
+        var fileWriter: FileWriter? = null
+        var csvPrinter: CSVPrinter? = null
+        val file = File("${context.filesDir.path}/export_dictionary.csv")
+        try {
+            file.createNewFile()
+            fileWriter = FileWriter(file)
+            csvPrinter = CSVPrinter(fileWriter, CSVFormat.DEFAULT)
+
+            for (word in words) {
+                val data = Arrays.asList(
+                        word.translation,
+                        word.word
+                )
+                csvPrinter.printRecord(data)
+            }
+        } catch (e: Exception) {
+            e.printStackTrace()
+            Log.e("ERROR", e.message)
+        } finally {
+            try {
+                fileWriter!!.flush()
```

Comment:
How will this work if the exception happened at the creation of the `fileWriter`?

Reply:
It won't. I removed the finally block and added the close and flush methods to the first try block.